src/Modules/flattenings.jl

```diff
@@ -0,0 +1,100 @@
+### Flattenings of (graded) modules
+
+function flatten(F::ModuleFP)
+  return flatten(base_ring(F))(F)
+end
+
+function (flat_map::RingFlattening)(F::ModuleFP{T}) where {T <: MPolyRingElem{<:Union{MPolyRingElem, MPolyQuoRingElem, 
+                                                                  MPolyQuoLocRingElem, MPolyLocRingElem}}}
+  if !haskey(flat_counterparts(flat_map), F)
+    F_flat, iso = change_base_ring(flat_map, F)
+    iso_inv = hom(F_flat, F, gens(F), inverse(flat_map))
+    set_attribute!(iso, :inverse, iso_inv)
+    set_attribute!(iso_inv, :inverse, iso)
+    flat_counterparts(flat_map)[F] = (F_flat, iso, iso_inv)
+  end
+    
+  F_flat, iso, iso_inv = flat_counterparts(flat_map)[F]::Tuple{<:ModuleFP, <:ModuleFPHom, <:ModuleFPHom}
+  return F_flat, iso, iso_inv
+end
+
+function flatten(
```

Comment:
I pursued the following approach: There is one user facing method `flatten` that can be called on any reasonable object. The user does not need to care about which flattening is being used and where it is cached, etc. Internally, a `RingFlattening` of `R` is created and stored in the ring `R`. This one is called by default in the `flatten` methods and it maintains all associated flattened objects like ideals, modules, etc. The internal user/programmer can or rather should use a specified flattening and not the cached one when coding. 

Does that make sense?